lib/src/facts/cache.cc

```diff
@@ -16,6 +16,22 @@ namespace boost_file = boost::filesystem;
 
 namespace facter { namespace facts { namespace cache {
 
+    void clean_cache(unordered_map<string, int64_t> facts_to_cache, string cache_location) {
+        boost_file::path cache_dir = boost_file::path(cache_location);
+        if (!boost_file::is_directory(cache_dir)) {
+            return;
+        }
+        for (boost_file::directory_iterator itr(cache_dir);
+                itr != boost_file::directory_iterator();
+                ++itr) {
+            boost_file::path cache_file = itr->path();
+            if (!facts_to_cache.count(cache_file.filename().string())) {
+                LOG_DEBUG("Deleting unused cache file {1}", cache_file.string());
+                boost_file::remove(cache_file);
```

Comment:
This could throw an exception if the directory is deleted between checking for existence and removing it, or if we don't have permission to delete it. I'd pass the error_code argument and LOG_DEBUG if there was an error, but otherwise continue.